// @datadog/browser-logs/cjs/domain/report/reportCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.startReportCollection = startReportCollection;
var browser_core_1 = require("@datadog/browser-core");
var isAuthorized_1 = require("../logger/isAuthorized");
var createErrorFieldFromRawError_1 = require("../createErrorFieldFromRawError");
function startReportCollection(configuration, lifeCycle) {
    var reportSubscription = (0, browser_core_1.initReportObservable)(configuration, configuration.forwardReports).subscribe(function (rawError) {
        var message = rawError.message;
        var error;
        var status = rawError.originalError.type === 'deprecation' ? isAuthorized_1.StatusType.warn : isAuthorized_1.StatusType.error;
        if (status === isAuthorized_1.StatusType.error) {
            error = (0, createErrorFieldFromRawError_1.createErrorFieldFromRawError)(rawError);
        }
        else if (rawError.stack) {
            message += " Found in ".concat((0, browser_core_1.getFileFromStackTraceString)(rawError.stack));
        }
        lifeCycle.notify(0 /* LifeCycleEventType.RAW_LOG_COLLECTED */, {
            rawLogsEvent: {
                date: (0, browser_core_1.timeStampNow)(),
                message: message,
                origin: browser_core_1.ErrorSource.REPORT,
                error: error,
                status: status,
            },
        });
    });
    return {
        stop: function () {
            reportSubscription.unsubscribe();
        },
    };
}
